pkg/compact/downsample/downsample.go

```diff
@@ -114,127 +112,24 @@ func Downsample(
 			resolution,
 		)
 		if err != nil {
-			return id, errors.Wrap(err, "downsample aggregate block")
+			return id, errors.Wrapf(err, "downsample aggregate block, series: %d", pall.At())
+		}
+		if err := streamedBlockWriter.AddSeries(&series{lset: lset, chunks: res}); err != nil {
+			return id, errors.Wrapf(err, "downsample aggregated block, series: %d", pall.At())
 		}
-		newb.addSeries(&series{lset: lset, chunks: res})
 	}
 	if pall.Err() != nil {
 		return id, errors.Wrap(pall.Err(), "iterate series set")
 	}
-	comp, err := tsdb.NewLeveledCompactor(nil, log.NewNopLogger(), []int64{rng}, NewPool())
-	if err != nil {
-		return id, errors.Wrap(err, "create compactor")
-	}
-	id, err = comp.Write(dir, newb, origMeta.MinTime, origMeta.MaxTime)
-	if err != nil {
-		return id, errors.Wrap(err, "compact head")
-	}
-	bdir := filepath.Join(dir, id.String())
-
-	var tmeta block.ThanosMeta
-	tmeta = origMeta.Thanos
-	tmeta.Source = block.CompactorSource
-	tmeta.Downsample.Resolution = resolution
 
-	_, err = block.InjectThanosMeta(logger, bdir, tmeta, &origMeta.BlockMeta)
+	id, err = streamedBlockWriter.Flush()
```

Comment:
probably, can inclose this into the Close method to write index with meta if there wasn't any error during the process and clean up resources then.

What do you think? @bwplotka

Reply:
Yea I was thinking about the same, but Close is usually ommitted by users -> error is almost never properly handled. 

Good write up about the problem is here: https://www.joeshaw.org/dont-defer-close-on-writable-files/

I feel like still doing finishing/important flush/writes in `Close` is a good pattern, so yea - can we move Flush to Close?

Reply:
I just realized, moving it to single method makes this overcomplicated, that's not only writer, so in case error happens in the time of chunk processing, it's not needed to flush index and meta data.

Reply:
It's tradeoff -> complication internally vs on conumer part. What is complicated, isn't it as easy as

```
func (w *writer) Close() error {
  if w.err == nil {
      // Do flush 
  }
  // Close
}
```

Reply:
Anyway I don't think that is major blocking decision, so up to you

Reply:
err could happen not only on writer side, that's the problem

i.e:
here https://github.com/improbable-eng/thanos/pull/529/files#diff-dc3c081796dd4ba1a1fdf96d58ab1f44R84 or here https://github.com/improbable-eng/thanos/pull/529/files#diff-dc3c081796dd4ba1a1fdf96d58ab1f44R115